File: LeetCode/Python/Medium/259-3Sum-Smaller/solution.py
class Solution:
    def threeSumSmaller(self, nums: List[int], target: int) -> int:
        result=0

        if not nums:
            return result

        n=len(nums)

        if n < 3:
            return result

        nums.sort()

        for index in range(0, n-2):
            desired_sum=target - nums[index]

            left=index + 1

            right=n - 1

            while left < right:
                current_sum=nums[left] + nums[right]

                if current_sum < desired_sum:
                    pairs=right - left
                    
                    result+=pairs

                    left+=1
                else:
                    right-=1

        return result

File: LeetCode/Rust/Medium/259-3Sum-Smaller/solution.rs
impl Solution
{
    pub fn three_sum_smaller(mut nums: Vec<i32>, target: i32) -> i32
    {
        let mut result: i32=0;

        if nums.is_empty()
        {
            return result;
        }

        let n: usize=nums.len();

        if n < 3
        {
            return result;
        }

        nums.sort();

        for index in 0..n-2
        {
            let mut left: usize=index + 1;

            let mut right: usize=n - 1;

            let desired_sum: i32=target - nums[index];

            while left < right
            {
                let current_sum: i32=nums[left] + nums[right];

                if current_sum < desired_sum
                {
                    let pairs: i32=(right - left) as i32;

                    result+=pairs;

                    left+=1;
                }
                else
                {
                    right-=1;
                }
            }
        }

        result
    }
}